Propose HP Dispatcher (#7)
Holo-Host/rfcs: 2019-11-03-hp-nginx/README.md

### X-Holo-Admin
HP Admin related calls have to be secured with the Signature-based authorization. It needs to be handled by HP-dispatcher because Holochain Conductor, which receives websocket traffic, does not have capability for authorization check. 

Authorization schema is `X-Holo-Admin-Signature` HTTP header, followed by Base64-encoded Ed25519 signature of the following payload:
```
{
  "method": ${verb}, // eg. "GET"
  "request": ${endpoint}, // eg "/api/v1/config"
  "body": ${body} // eg. "{\"name\": \"My HoloPort Name\"}"
}
```

Example: `X-Holo-Admin-Signature: EGeYSAmjxp1kNBzXAR2kv7m3BNxyREZnVwSfh3FX7Ew`

### none for hosting

Authorization for hosting related traffic is still an open question. Can any user open a websocket connection with Conductor? How can Conductor handle concurrency in requests?

## Transport Layer Reservations

The `name` in `<ip>:<port> "<name>"` corrosponds to the port number in that it matches the alpha/numeric mapping
of English telephone keypads (eg. https://phonespell.org).

### `0.0.0.0:443` Reverse Proxy

**Supported protocols**
- HTTPS
- WebSocket Secure

### `localhost:4656 "holo"` Envoy

**Supported protocols**
- HTTP
- WebSocket

> ### `localhost:23646 "admin"` Admin Server

Reviewer: Admin server will actually have to be a Unix socket, for security reasons. It would be best to switch to Unix sockets wherever possible.

# HP Dispatcher

## The purpose

HoloPort is running multiple services that need to be available to end user. Traffic is directed to the HoloPort by holo-router over the ZeroTier network. HoloPort needs to have a 443 inbound port open on the ZeroTier network interface. On that port (443) HP-dispatcher (eg. implemented via nginx) is listening.

## TLS

HP-dispatcher terminates https traffic. TLS Certificate installation is handled by nixOS and Let's Encrypt SSL service.

## Routes

HP-dispatcher needs to handle following connections:

| Route         | Authorization                   | Type | Destination                  | Purpose                                      |
| -----         | -------------                   | ---- | -----------                  | -------                                      |

> | `/`           | [`rate-limit`](#rate-limit)     | http | 302 redirect to              | Redirect to `/admin/`                        |

Reviewer: Not sure why this is a redirect instead of serving `/admin` directly.

Author: To stick to semantics in the url: "now you are served THIS UI"

> | `/admin/`     | [`rate-limit`](#rate-limit)     | http | HP Admin static files folder | Serve static files of HP Admin               |

Reviewer: ```suggestion
```

Author: same as above - semantics of the url

| `/holofuel/`  | [`rate-limit`](#rate-limit)     | http | HoloFuel static files folder | Serve static files of HoloFuel               |
| `/api/v1/`    | [`X-Holo-Admin`](#X-Holo-Admin) | http | port 23646                   | Proxy to HP Admin Server                     |
| `/api/v1/ws/` | [`X-Holo-Admin`](#X-Holo-Admin) | ws   | port 42233                   | Proxy to Conductor Admin Instances           |

> | `/hosting/`   | none*                           | ws   | port 4656                    | Proxy to Envoy service handling Holo-hosting |

Reviewer: ```suggestion
| `/v1/hosting/`   | none*                           | ws   | port 4656                    | Proxy to Envoy service handling Holo-hosting |
```

> # HP Nginx

Reviewer: Name couples specific implementation with the interface. Could we go with "Dispatcher" or something along those lines?

## The purpose

> HoloPort is running multiple services that need to be available to end user. Traffic is directed to the HoloPort by holo-router over the ZeroTier network. HoloPort needs to have a 443 inbound port open on the ZeroTier network interface. On that port (443) HP-nginx is listening.

Reviewer: For consistency:

```suggestion
HoloPort is running multiple services that need to be available to end user. Traffic is directed to the HoloPort by holo-router over the ZeroTier network. HoloPort needs to have a 443 inbound port open on the ZeroTier network interface. On that port (443) HP Nginx is listening.
```

## TLS

> Nginx terminates https traffic. TLS Certificate installation is handled by nixOS and Let's Encrypt SSL service.

Reviewer: Nitpicks:

```suggestion
Nginx terminates HTTPS traffic. TLS certificate installation is handled by NixOS and Let's Encrypt service.
```

## Routes

Nginx needs to handle following connections:

| Route | Authorization | Type | Destination | Purpose |
| ----- | ------------- | ---- | ----------- | ------- |

> | `/` | [`rate-limit`](#rate-limit) | http | 302 redirect to  | Serve static files of HP Admin |

Reviewer: Why not serve admin interface from `/` in the first place?

| `/admin/`   | [`rate-limit`](#rate-limit) | http | HP Admin static files folder | Serve static files of HP Admin |
| `/holofuel/` | [`rate-limit`](#rate-limit) | http | HoloFuel static files folder | Serve static files of HoloFuel |
| `/api/v1/` | [`X-Holo-Admin`](#X-Holo-Admin) | http | port xxx | Redirect to HP Admin Server |
| `/api/v1/ws/` | [`X-Holo-Admin`](#X-Holo-Admin) | ws | port xxx | Redirect to conductor running HoloFuel and HoloHosting app |

> | `/hosting/` | none* | ws | port xxx | Redirect to envoy service handling Holo-hosting |

Reviewer: My understanding of the plan, that I shared today on standup and previously, was that we'll control installed hApps using Nix. TBH, I don't know how to integrate hApp Store into HoloPortOS any other way. That would mean, it's done completely through HP Admin API `holoportos` subfield.

Return `404 Not Found` otherwise.

## Authorization

Depending on the service there are various ways for authorizing traffic.

### rate-limit

Static assets server will have to have rate limit on number of requests per one IP.
```

> limit_req_zone $binary_remote_addr zone=zone1:1m rate=2r/s;

Reviewer: Could you explain what rate limit policy behind this config?

location / {
    limit_req zone=zone1 burst=30;
}
```

> ### X-Holo-Admin

Reviewer: For consistency: 
```suggestion
### X-Holo-Admin-Signature
```

Reviewer: I wonder if this is still a good header name for this. `x-hp-nginx-signature` would be more descriptive?

Reviewer: I wonder if this is still a good header name for this. `X-HP-Dispatcher-Signature` would be more descriptive?

Reviewer: ```suggestion
### X-Hpos-Admin-Signature
```

HP Admin related calls have to be secured with the Signature-based authorization. It needs to be calculated in HP-nginx because Holochain Conductor, which receives websocket traffic, does not have capability for authorization check. 

> Authorization schema is `X-Holo-Admin-Signature` HTTP header, followed by Base64-encoded Ed25519 signature. Payload signed is ???

Reviewer: Probably, this would have to be an ad-hoc HTTP request representation in JSON, except for signature header. Something like:

```json
{
  "method": "PUT",
  "request": "/api/v1/config",
  "headers": {
    "content-type": ["application/json"]
  },
  "body": "{\"name\": \"My HoloPort Name\"}"
}
```

# HP Dispatcher

## The purpose

> HoloPort is running multiple services that need to be available to end user. Traffic is directed to the HoloPort by holo-router over the ZeroTier network. HoloPort needs to have a 443 inbound port open on the ZeroTier network interface. On that port (443) HP-dispatcher (eg. implemented via nginx) is listening.

Reviewer: ```suggestion
HoloPort is running multiple services that need to be available to end user. Traffic is directed to the HoloPort by Holo Router over the ZeroTier network. HoloPort needs to have inbound 443 port open on the ZeroTier network interface. On that port (443) HP Dispatcher (e. g. implemented via Nginx) is listening.
```

## TLS

HP-dispatcher terminates https traffic. TLS Certificate installation is handled by nixOS and Let's Encrypt SSL service.

## Routes

> HP-dispatcher needs to handle following connections:

Reviewer: ```suggestion
HP Dispatcher needs to handle following connections:
```

| Route | Authorization | Type | Destination | Purpose |
| ----- | ------------- | ---- | ----------- | ------- |
| `/` | [`rate-limit`](#rate-limit) | http | 302 redirect to  | Serve static files of HP Admin |
| `/admin/`   | [`rate-limit`](#rate-limit) | http | HP Admin static files folder | Serve static files of HP Admin |
| `/holofuel/` | [`rate-limit`](#rate-limit) | http | HoloFuel static files folder | Serve static files of HoloFuel |
| `/api/v1/` | [`X-Holo-Admin`](#X-Holo-Admin) | http | port xxx | Redirect to HP Admin Server |
| `/api/v1/ws/` | [`X-Holo-Admin`](#X-Holo-Admin) | ws | port xxx | Redirect to conductor running HoloFuel and HoloHosting app |
| `/hosting/` | none* | ws | port xxx | Redirect to envoy service handling Holo-hosting |

Return `404 Not Found` otherwise.

> ## Authorization

Reviewer: This section should probably allude to the fact that, since we don't sign HTTP headers, WebSocket MITM is possible until full-featured light client is implemented.

Depending on the service there are various ways for authorizing traffic.

> ### rate-limit

Reviewer: ```suggestion
### Rate limit
```

Static assets server will have to have rate limit on number of requests per one IP.
```
limit_req_zone $binary_remote_addr zone=zone1:1m rate=2r/s;

location / {
    limit_req zone=zone1 burst=30;
}
```
This is a measure against bandwidth depletion due to unwanted serving of static files from HoloPort, as the are meant to be served to HoloPort Admin alone.

### X-Holo-Admin

> HP Admin related calls have to be secured with the Signature-based authorization. It needs to be handled by HP-dispatcher because Holochain Conductor, which receives websocket traffic, does not have capability for authorization check. 

Reviewer: ```suggestion
HP Admin related calls have to be secured with the Signature-based authorization. It needs to be handled by HP Dispatcher because Holochain Conductor, which receives WebSocket traffic, does not have capability for authorization check. 
```
